11789 e2o launch args

erik-ciq opened the pull request:
**Resolves issue [11789](https://chartiq.kanbanize.com/ctrl_board/18/cards/11789/details)**

**Description of change**
Add example of customizing launchE2O and launch args into gulpfile-extensions.js. Add envOrArg to taskMethods.

**Description of testing**
App still launches, e2o included the flags specified in gulpfile-extensions `launchE2o`

NinjaManatee commented on gulpfile.js:
```diff
@@ -437,6 +437,10 @@
 			logToTerminal.apply(this, arguments);
 		},
 
+		envOrArg: (...args) => {
```
Is it better to do this or just move the implementation of envOrArg here?

carterbs replied:
@erik-ciq please confirm with andy that he is satisfied and resolve this conversation.

NinjaManatee commented on gulpfile-extensions-example.js:
```diff
@@ -14,7 +14,9 @@ module.exports = taskMethods => {
 
     const gulp = require("gulp");
-
+	const ON_DEATH = require("death")({ debug: false });
```
Whitespace is off

carterbs replied:
@erik-ciq please get with andy/confirm this comment is resolved.

erik-ciq commented:
@NinjaManatee there is a lot of copy and pasted code, I'm also not a fan.

One alternate approach would be to do this via CLI flags / environment variables. So when a user launches finsemble-seed they could do something like: `npm run dev --auth-server-whitelist='*'`, or just set `auth-server-whitelist` as an environment variable. 

Another alternate approach is to create a new function like `launchE2OArgs` that returns an object of arguments to use when launching e2o. If we were to do that, we could eliminate almost all of the copy and paste code. 

I would be happy to update this PR with a better approach. I think it would be worth it.

carterbs commented:
QA notes....much of this may be due to me not knowing what I'm doing.

1. If I put anything after the manifest argument, the whole thing dies.
2. When it dies, it errors out because errorOutColor isn't a function.
3. If I put a nonsense param in the middle (`--twiddle yourThumbs`), the output is this:
```

0 -> C:\Users\BradC\git\finsemble-seed\node_modules\electron\dist\electron.exe                                                                                                                                                                                                                                                                                              1 -> index.js                                                                                                                                                                                                                                                                                                                                                               2 -> --auth-server-whitelist='*'                                                                                                                                                                                                                                                                                                                                            3 -> --remote-debugging-port=9090                                                                                                                                                                                                                                                                                                                                           4 -> NaN                                                                                                                                                                                                                                                                                                                                                                    5 -> --manifest                                                                                                                                                                                                                                                                                                                                                             6 -> http://localhost:3375/configs/openfin/manifest-local.json
--
```

I don't know if this is a bug or because it's trying to parse --twiddle as the port for the remote debugging port.

carterbs commented:
@erik-ciq take a look at my comments above. errorColorOut is the only thing that should probaby actually be fixed? Not sure if the other stuff is a bug.

erik-ciq commented:
@BradleyCarter went ahead and fixed the issue with errorColorOut, also I tested the `--twiddle yourThumbs` in the middle, that one launched for me once I fixed the errorColorOut issue. Everything seems to be working now.

erik-ciq commented on gulpfile-extensions-example.js:
```diff
@@ -27,6 +29,69 @@ module.exports = taskMethods => {
+		});
+
+		electronProcess.on("close", function (code) {
+			console.log("child process exited with code " + code);
```
I went ahead and updated the console log and removed the unnecessary commented out code.

meredith-ciq commented on gulpfile-extensions-example.js:
```diff
@@ -27,6 +33,65 @@ module.exports = taskMethods => {
+					console.error(errorOutColor(err));
+				}
+
+				if (watchClose) watchClose();
```
watchClose is throwing a not defined exception when I ctrl+c to kill finsemble. I'm not sure about regular shutdown because of the existing childprocess error, but this variable is not defined in this file.

erik-ciq commented:
Went ahead and updated the path to `electron.exe`, made sure that the `envOrArg.apply` is being returned properly, defined `watchClose` in `gulpfile-extensions-example.js`, and commented out `launchE2O` in `gulpfile-extensions-example.js`. Should be good now.

meredith-ciq commented on gulpfile.js:
```diff
@@ -432,9 +432,7 @@
 			logToTerminal.apply(this, arguments);
```
@erik-ciq Can you make the same change as envOrArg here? It's only logging [object, object]. Otherwise these changes are great, thanks!

erik-ciq replied:
@meredith-ciq went ahead and updated that.

meredith-ciq commented on gulpfile.js:
```diff
@@ -433,9 +428,9 @@
-		logToTerminal: () => {
-			logToTerminal.apply(this, arguments);
-		},
+		logToTerminal: () => logToTerminal.apply(this, arguments),
```
@erik-ciq Sorry I just assumed the problem was the same as envOrArg without testing, but it's not. arguments is a reserved word with a lot of junk in it, but I don't think our message is inside. It needs to pass in args like envOrArg.
logToTerminal: (...args) => logToTerminal.apply(this, args),

erik-ciq replied:
That makes sense. Updated to use args instead of arguments.
